src/common/transformations/src/transformations/mlir/op/matmul.cpp

```diff
@@ -0,0 +1,67 @@
+// Copyright (C) 2018-2024 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#include "mlir/Dialect/Tensor/IR/Tensor.h"
+#include "mlir/Dialect/Linalg/Passes.h"
+
+#include <openvino/op/matmul.hpp>
+#include "openvino/pass/pattern/op/wrap_type.hpp"
+
+#include "matmul.hpp"
+#include "../convert_common.hpp"
+
+
+namespace {
+
+using namespace ov::mlir;
+
+struct ConvertMatMul {
+    void operator()(ConversionContext& context, NodePtr node) {
+        auto loc = createLocation(context.context, node);
+        auto& builder = context.builder();
+        // TODO: Support broadcasts
+        const auto inputs = context.getInputs(node);
+        const auto ov_output_element_type = node->get_output_element_type(0);
+        const auto ov_output_shape = node->get_output_partial_shape(0);
+        auto outType = importTensor(context.context, ov_output_shape, ov_output_element_type); // Instead of this (WRONG): cast<mlir::ShapedType>(inputs[0].getType());
+        // Named binary ops directly overwrite data in `outs` buffer so, there is no need to provide non-empty
+        // destination at the tensor-level.
+        // Use `tensor.empty` to avoid temporary buffer allocation and memcpy after bufferization.
```

Comment:
That's true for named element-wise operations but not for `linalg.matmul` which uses `outs` as initial value for the accumulation (`outs += lhs * rhs`).
If graph doesn't provide any output value then we should zero fill the empty tensor:
```mlir
%zero = arith.constant 0
%e = tensor.empty
%fill = linalg.fill ins(%zero) outs(%e)
%res = linalg.matmul ins(%A, %B) outs(%fill)
```

Reply:
That's probably why I have random results :)